### PR TITLE
Improve vesting discovery and error messages

### PR DESCRIPTION
**Pull request by sebastinez**
When loading a ETH address as vesting contract we show a cryptic error
message, which non tech-savy end users don't understand, 8c847b0
changes it to something more human readable.

We allow users to enter and load a contract that probably isn't a
vesting contract, we show this now by showing an error message.

The `getInfo` function in the `vesting` folder is very prone to possible
errors, and benefits from some parallelization between all the method
calls, and some error handling.

We allow hereby to enter a contract address in search and navigate to the
contract profile page where we provide a button to display the detailed vesting
information if available.

On the same profile page, the user has the possibility to withdraw their withdrawable balance
also the balance is being increased on every new block.

### CODE REVIEW

**Comment by sebastinez** [comment]
Here a video of some of the improvements:

Show a error message when trying to load a contract that probably isn't a vesting contract
Bildschirm­foto 2022-12-21 um 18 25 23

Show a link to the vesting information on a contract profile page if available
We preload the vestingInfo so the loading time is instant.
Bildschirm­foto 2022-12-21 um 18 27 43

If we try to lookup a vesting contract that doesn't exist with `/vesting/<contract address>` we show a `NotFound` component with some text for the user
Bildschirm­foto 2022-12-21 um 18 29 59

File: public/typography.css
Comment (rudolfs) [inline public/typography.css:140]: We're just using this in a single place, maybe we can just inline it there?

File: src/App.svelte
Comment (rudolfs) [inline src/App.svelte:105]: Do we even want to keep the special vesting route now that we can just use the search bar?

Comment (sebastinez) [reply src/App.svelte:105]: Yeah I wanted to remove it initially but some core contributors received mails with the /vesting route, to enter/load their contract.
So didn't want to break that.
Maybe at some point in the future?

**Comment by rudolfs** [comment]
Why do we have the vesting toggle button? Could we just show in the badge that it's a vesting contract?

Screenshot 2022-12-22 at 18 02 39

**Comment by rudolfs** [comment]
Gonna tag @brandonhaslegs here, maybe he has some ideas on the design. The interaction feels a bit weird to me, wonder if there's a better way? Also the link to Etherscan could be nicer.

https://user-images.githubusercontent.com/158411/209188684-db0bebf3-c91d-4249-8e0f-be34bbf1d52f.mov



Screenshot 2022-12-22 at 18 04 28

**Comment by sebastinez** [comment]
> Why do we have the vesting toggle button? Could we just show in the badge that it's a vesting contract?

Yeah this was the initial way I thought of doing it, but I didn't liked the amount of lines, showing up, so the idea of the vesting toggle button is to hide all that information behind a toggle.

@brandonhaslegs do you have any idea here maybe on how to make this nicer, I don't think we should show all the vesting information from the get go, since it uses lot of space.

File: public/typography.css
Comment (sebastinez) [reply public/typography.css:140]: Yeah I wanted to keep the line of doing it, and eventually with more error handling, we could start using something like this. But for now we can do it inline

File: src/App.svelte
Comment (rudolfs) [reply src/App.svelte:105]: Or we could just show instructions on that page that say: please paste your vesting contract into the search bar at the top?

File: src/base/vesting/vesting.ts
Comment (rudolfs) [inline src/base/vesting/vesting.ts:58]: We shouldn't cast errors to `any`, by default it's `unknown`. Further down we should check that the error actually has a `.reason` attribute.

File: src/base/vesting/vesting.ts
Comment (rudolfs) [inline src/base/vesting/vesting.ts:59]: `console.warn`?

File: src/base/vesting/vesting.ts
Comment (rudolfs) [inline src/base/vesting/vesting.ts:80]: Shouldn't we handle the `undefined` case here instead of asserting that it's always a `string`?

Screenshot 2022-12-23 at 09 59 12

Comment (rudolfs) [inline src/base/vesting/vesting.ts:80]: Could we move this out of the try block and just do:
```
token = await contract.token();
if(!token) {
return undefined;
}
```

Instead of the type assertion? I would find that easier to read.

File: src/base/vesting/vesting.ts
Comment (rudolfs) [inline src/base/vesting/vesting.ts:125]:
```suggestion
export const parsedVestingPeriods = (...timestamps: string[]): string => {
```

File: src/App.svelte
Comment (sebastinez) [reply src/App.svelte:105]: Yeah that doesn't sound like good UX.
IMO either:
- we have a page like currently where we only show vesting information under a specific route.
- or we add it to a profile page like this PR wants to do it, and:
  -  show it in some nice condensed graphical easy to understand way with graphs maybe?
  -  or we hide it with a dropdown to avoid having lots of lines of information.

File: src/base/vesting/vesting.ts
Comment (sebastinez) [reply src/base/vesting/vesting.ts:80]: Hmm the reason I left it inside the `try` block is to catch any errors that could happen when fetching it.
I think I could leave it inside the `try` block but do the `if(!token)` guard to avoid the casting to `string`

Comment (sebastinez) [reply src/base/vesting/vesting.ts:80]: Something like
```js
  let token: string | undefined = undefined;

  try {
    token = await contract.token();
    if (!token) {
      return undefined;
    }
```

**Comment by brandonhaslegs** [comment]
This whole page is about vesting right? I don’t think you need to hide anything behind a button. What else is there to do on this page?

**Comment by sebastinez** [comment]
> This whole page is about vesting right? I don’t think you need to hide anything behind a button. What else is there to do on this page?

The page is the profile page of a ethereum address, so there will be a listing of projects corresponding to a user or eventual org, and personal information like seed maybe balances for multi sig and eventually other stuff.

File: public/typography.css
Comment (sebastinez) [reply public/typography.css]: 634126e

File: src/base/vesting/vesting.ts
Comment (sebastinez) [reply src/base/vesting/vesting.ts]: 28746cd should improve it

File: src/base/vesting/vesting.ts
Comment (sebastinez) [reply src/base/vesting/vesting.ts]: 28746cd

**Comment by brandonhaslegs** [comment]
Or instead of just showing a button, can you show some value there?

So instead of:

**Vesting** [Show details]

you could indead show a value with a [View more] button next to it

**Vesting Beneficiary** 👤rudolfs.eth [View more]

And the “view more” button would then expand all the vesting info.

**Comment by brandonhaslegs** [comment]
Things I think should change:
- The [back] button is weird. It should just timeout.
- The loading indicator looked better as the three-dot that’s everywhere else in the app
- The eth address after “Waiting for confirmation” should be truncated just like the address is truncated with the “Tokens have been withdrawn” success state. I made a separate issue for truncating all eth addresses here #555

**Comment by brandonhaslegs** [comment]
I am also wondering if this withdrawl flow should be more “generic” like just exist in a modal, just in case you have to withdrawl from a different UI at some point. Then we could repurpose the UI.

**Comment by sebastinez** [comment]
Thanks @brandonhaslegs for the feedback! Really appreciated!

> Or instead of just showing a button, can you show some value there?

That's a great idea, I'll think about some values we can show together with a [View more] toggle

> The [back] button is weird. It should just timeout.

Ok I'll try some other timeout durations, maybe we could also do something like when moving your mouse out of the textarea we timeout after some seconds, kinda like a debounce after user interaction? 

> The loading indicator looked better as the three-dot that’s everywhere else in the app

Got it

> The eth address after “Waiting for confirmation” should be truncated just like the address is truncated with the “Tokens have been withdrawn” success state.

The address shown on "Waiting for confirmation" is the transaction hash of the withdrawal, I thought it would be nice to give users this information in case something goes bad.
The transaction hash is trimmed to it's first 20 characters of 66 in total, I took the inspiration from etherscan but we can do it differently or don't show the tx hash at all 🤷 
Bildschirm­foto 2022-12-23 um 22 40 46

> I am also wondering if this withdrawl flow should be more “generic” (for example just exist in a modal flow) just in case you have to withdrawl from a different UI at some point. Then we could repurpose the withdrawl UI. This would have to be designed of course, but we’ve designed many withdrawl flows at this point, I’m sure we could throw something together. Maybe this is something better to do in the future.

Really like the idea of a generic withdrawal component modal that can be shown at any point when there is some withdraw happening. Maybe that's something we can take into account in the modal refactor at #557

**Comment by sebastinez** [comment]
Okay I addressed your feedback @brandonhaslegs and I'm open to review the etherscan icon https://github.com/radicle-dev/radicle-interface/pull/552#issuecomment-1363110256

I would like to get this PR closed before heartwood comes out.

**Comment by rudolfs** [comment]
> I would like to get this PR closed before heartwood comes out.

You mean you want to merge it into master?

**Comment by sebastinez** [comment]
> You mean you want to merge it into master?

Yeah exactly 👍

**Comment by brandonhaslegs** [comment]
Here is a new etherscan icon

etherscan

**Comment by brandonhaslegs** [comment]
Can you give me a URL where I can test this?

**Comment by sebastinez** [comment]
> @sebastinez Can you give me a URL where I can test this?

Yeah sure, let me update the icon and I'll deploy it somewhere

**Comment by brandonhaslegs** [comment]
Ah sorry I built it just fine from terminal, i just don’t know where to go to find this page

**Comment by sebastinez** [comment]
> Ah sorry I built it just fine from terminal, i just don’t know where to go to find this page

Ok I just added a mocking function to test it more easily.
Pull the latest changes and try out this url http://localhost:3000/0x27BCA0692e13C122E6Fc105b3974B5df7246D464

**Comment by sebastinez** [comment]
Or it can also be previewed here https://tubular-taiyaki-bf2c09.netlify.app/0x27BCA0692e13C122E6Fc105b3974B5df7246D464

**Comment by rudolfs** [comment]
> I am also wondering if this withdrawl flow should be more “generic” (for example just exist in a modal flow) just in case you have to withdrawl from a different UI at some point. Then we could repurpose the withdrawl UI. This would have to be designed of course, but we’ve designed many withdrawl flows at this point, I’m sure we could throw something together. Maybe this is something better to do in the future.

@brandonhaslegs this is how it is implemented now actually and this PR changes it to not be in a modal.

See: https://app.radicle.xyz/vesting

Here are some sample Görli vesting contracts:
```
0x13b2Fc1f601Fb72b86BFAB59090f22bB6E73005A
0xAADcbc69f955523B0ff0A271229961E950538EbE
0x27BCA0692e13C122E6Fc105b3974B5df7246D464
0x9c882463B02221b0558112dec05F60D5B3D99b6a
```


Original:

Screenshot 2022-12-28 at 15 43 07

Screenshot 2022-12-28 at 15 43 43

**Comment by brandonhaslegs** [comment]
Well why dont we keep this with a few changes?

**Comment by sebastinez** [comment]
> Well why dont we keep this with a few changes?

The idea of implementing the whole vesting information into a Profile page came up during a discussion here: https://github.com/radicle-dev/radicle-interface/issues/526#issuecomment-1338144182

I think if we would like to keep everything as is, we could just redirect searches for vesting contracts to the modal view.
And improve the withdraw process, and error handling in the vesting modal. 🤔

**Comment by sebastinez** [comment]
Ok I think this PR is still better than the current status quo.. but if anyone wants to take another stab at designing this or has already something, we can sync maybe over Discord and bring this PR over the finish line.

**Comment by cloudhead** [comment]
Feedback here:
image

**Comment by cloudhead** [comment]
Furthermore:

*withdraw* button should not show if there's nothing to withdraw or if you're not the beneficiary.
If you're not connected, it should be disabled.

**Comment by sebastinez** [comment]
Ok addressed the feedback by @cloudhead.

Here
Bildschirm­foto 2023-01-09 um 15 37 33

And here some videos of the withdrawal flow.
One receives an error during the transaction wait, and one completes sucessfully.

https://user-images.githubusercontent.com/7912302/211340818-f4d60593-7c62-41ec-9829-52be996a869c.mov

https://user-images.githubusercontent.com/7912302/211340828-392d434c-88f8-4541-825f-268085471280.mov

**Comment by rudolfs** [comment]
The etherscan icon doesn't match the rest of the social-media link icons:
Screenshot 2023-01-09 at 16 30 28

It also doesn't seem to be centered in the bounding box.
Screenshot 2023-01-09 at 16 30 55

Here are all our other icons:
Screenshot 2023-01-09 at 16 37 50

**Comment by rudolfs** [comment]
It's possible to submit the input via the <kbd>enter</kbd> key even when the button is disabled:
Screenshot 2023-01-09 at 16 42 43

We should not allow submit on keypress when the form isn't valid.

When there is no vesting contract, there are no errors in the console. We should adjust the error message to just "No vesting account found under this address."
Screenshot 2023-01-09 at 16 45 08

**Comment by rudolfs** [comment]
Is there a reason why this address is clickable? We are already on the profile page of it, no?
Screenshot 2023-01-09 at 16 49 37

**Comment by cloudhead** [comment]
Looking better! A couple more things:

* usually we show errors using the error modal which is all red; we don't mix red error text in blue modals
* In the success case, we should show the withdrawal address inline, eg. "tokens withdrawn to [icon] xyz - abc"

**Comment by rudolfs** [comment]
We should make the gap between the buttons a bit wider, like 1rem?
Screenshot 2023-01-09 at 16 54 10

**Comment by rudolfs** [comment]
When funds have been withdrawn and I click on the address in the modal, I get an error:

Screenshot 2023-01-09 at 16 56 43

Screenshot 2023-01-09 at 16 56 46

**Comment by cloudhead** [comment]
> Is there a reason why this address is clickable? We are already on the profile page of it, no?

I think that's a separate issue no? ie. it affects all profiles.

**Comment by cloudhead** [comment]
The etherscan icon is just bigger than the other icons for some reason.

**Comment by brandonhaslegs** [comment]
Yes! I recently redesigned our icon set (refined them, made them more streamlined). Here are the Github and Twitter ones. And the 🔗  one too!
github
link
twitter

**Comment by sebastinez** [comment]
Ok did update the Icons.

Bildschirm­foto 2023-01-09 um 22 07 17

Now etherscan surely seems more like the others.
@brandonhaslegs if you have other icons, for more streamlining please lmk and I'll do that in another PR
image
I think there are some others that could be improved

**Comment by sebastinez** [comment]
> It's possible to submit the input via the enter key even when the button is disabled: Screenshot 2023-01-09 at 16 42 43

Good catch! 💪

**Comment by sebastinez** [comment]
Ok addressed with the latest commit:

- the `ErrorModal` usage
- the withdrawal not working 😓 thanks for the catch @rudolfs!
- the error when clicking on the address where to tokens where withdrawn to.
- the ability to click enter and to trigger the `on:submit` of a `TextInput` component even if the input isn't valid.
- modified the error message, removing the dev console part.

A video as reference
https://user-images.githubusercontent.com/7912302/211507254-af37d79c-5c55-49b9-8a8c-9a1a88051e3f.mov

**Comment by cloudhead** [comment]
Thanks, looking good, just one thing left: spacing/alignment here is off:
image

I also wonder, what happens if you close the modal while it's confirming?

**Comment by sebastinez** [comment]
> Thanks, looking good, just one thing left: spacing/alignment here is off: image
> 
> I also wonder, what happens if you close the modal while it's confirming?

If you close the modal while confirming, after having signed the tx, the tx goes out either way and you loose eventually feedback if something goes wrong.
We could eventually just show the `Close` button once the funds have been withdrawn 🤔 but I wanted to give users the ability to close the modal also during the tx.

**Comment by sebastinez** [comment]
> Thanks, looking good, just one thing left: spacing/alignment here is off

@cloudhead Ok I fixed this by removing the `Address` component here, and do it as we do in the registration flow, where we just show the formatted address.

Bildschirm­foto 2023-01-10 um 10 34 25

**Comment by rudolfs** [comment]
The `withdrawn` amount doesn't update automatically after the tx is confirmed. It only updates after a page refresh.

File: src/lib/utils.ts
Comment (rudolfs) [inline src/lib/utils.ts:299]: We should rename the `addr` attribute to `addrOrTx`.

File: src/lib/utils.ts
Comment (rudolfs) [inline src/lib/utils.ts:323]: Were using this only in one place, could we inline it for now?

Comment (sebastinez) [reply src/lib/utils.ts]: 1e4ebf0

File: src/lib/utils.ts
Comment (sebastinez) [reply src/lib/utils.ts]: 1e4ebf0

**Comment by sebastinez** [comment]
> The `withdrawn` amount doesn't update automatically after the tx is confirmed. It only updates after a page refresh.

1e4ebf0

File: src/lib/vesting.ts
Comment (rudolfs) [inline src/lib/vesting.ts:108]: Can we log this error with `console.warn`?

File: src/lib/vesting.ts
Comment (rudolfs) [inline src/lib/vesting.ts:138]: Could we use a normal function definition instead of const where possible?

Comment (sebastinez) [reply src/lib/vesting.ts]: 16ead15

File: src/lib/vesting.ts
Comment (sebastinez) [reply src/lib/vesting.ts:109]: 16ead15

File: src/views/vesting/Withdraw.svelte
Comment (rudolfs) [inline src/views/vesting/Withdraw.svelte:48]: Could we use the ellipsis character … instead of `...` everywhere?

Comment (sebastinez) [reply src/views/vesting/Withdraw.svelte]: 5d2cbd3

**Review by rudolfs — APPROVED** [review]
👏 🚀

**Pull request merged**
